### PR TITLE
Limiting network traffic rate for uploading stage.

### PR DESCRIPTION
**Pull request by MikhailBurdukov**

### CODE REVIEW

**Review by aalexfvk — CHANGES_REQUESTED** [review]

File: ch_backup/util.py [hunk 1]
```diff
@@ -371,6 +371,38 @@ def dataclass_from_dict(type_: Type[T], data: dict) -> T:
...
+        if self._limit_per_sec == 0:
+            return True
+        current_time = self._get_time_func()
+        lapse = current_time - self._bucket_last_update
+        self._bucket_tokens = min(
+            self._limit_per_sec, self._bucket_tokens + lapse * self._limit_per_sec
+        )
+        if self._bucket_tokens >= tokens:
+            self._bucket_tokens -= tokens
+            self._bucket_last_update = current_time
```
Comment (aalexfvk) [inline ch_backup/util.py:400]: I think `last_update` must be updated after every replenish

File: ch_backup/util.py [hunk 2]
```diff
@@ -371,6 +371,38 @@ def dataclass_from_dict(type_: Type[T], data: dict) -> T:
     return type_(**{k: v for k, v in data.items() if k in class_fields})  # type: ignore[call-arg]
 
 
+class RateLimiter:
```
Comment (aalexfvk) [inline ch_backup/util.py:374]: Let's move this class into a separate import module from `util.py`

File: tests/integration/features/traffic_limiting.feature [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  Background:
+    Given default configuration
+    And a working s3
+    And a working zookeeper on zookeeper01
+    And a working clickhouse on clickhouse01
+    And clickhouse on clickhouse01 has test schema
+    And clickhouse01 has test clickhouse data test1
+    And we have executed queries on clickhouse01
+    """
+    CREATE USER test_user IDENTIFIED WITH plaintext_password BY 'password';
```
Comment (aalexfvk) [inline tests/integration/features/traffic_limiting.feature:12]: Let's test only data table without ACl entites. In my opinion they are complicating test. SImplify as far as possible

File: tests/integration/features/traffic_limiting.feature [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    When we restore clickhouse backup #0 to clickhouse02
+    And we restart clickhouse on clickhouse02
+    Given a working clickhouse on clickhouse02
+    Then we got same clickhouse data at clickhouse01 clickhouse02
+
+  Scenario: Test restore sources setted traffic rate limit.
+   Given ch-backup configuration on clickhouse01
+    """
+    storage:
+        uploading_traffic_limit_retry_time: 33554432
```
Comment (aalexfvk) [inline tests/integration/features/traffic_limiting.feature:35]: We can use `Examples:` clause to parametrize these both tests to one. Have a look at `deduplication.feature` for example.

File: tests/unit/test_rate_limiter.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+    return reduce(lambda pipeline, stage: pipeline | stage, stages)
+
+
+@given(
+    content_size=st.integers(16, 10000),
+    chunk_size=st.just(16),
+    buffer_size=st.just(128),
+)
+def test_traffic_rate_limiting(content_size, chunk_size, buffer_size):
```
Comment (aalexfvk) [inline tests/unit/test_rate_limiter.py:91]: Let's write comprehensive unit tests only for `RateLimiter` class. 
Testing on pipeline level looks cumbersome and overloaded with irrilevant details.

File: ch_backup/config.py [hunk 6]
```diff
@@ -101,6 +101,11 @@ def _as_seconds(t: str) -> int:
         "uploading_threads": 4,
         # The maximum number of objects the stage's input queue can hold simultaneously, `0`is unbounded
         "queue_size": 10,
+        # Upper bound of network loading per second on the uploading stage. (bytes/sec)
```
[the remark this answers is not in the document]
Comment (MikhailBurdukov) [reply ch_backup/config.py:106]: Fixed

File: ch_backup/storage/async_pipeline/pipeline_builder.py [hunk 7]
```diff
@@ -134,7 +135,8 @@ def build_uploading_stage(
         buffer_size = stage_config["buffer_size"]
         chunk_size = stage_config["chunk_size"]
         queue_size = stage_config["queue_size"]
-
+        traffic_limit_per_sec = stage_config["uploading_traffic_limit"]
```
[the remark this answers is not in the document]
Comment (MikhailBurdukov) [reply ch_backup/storage/async_pipeline/pipeline_builder.py]: Fixed

**Comment by MikhailBurdukov** [comment]
@aalexfvk, Fixed all requested changes in the last commit.

**Review by aalexfvk — APPROVED** [review]

**Pull request merged**